src/charm.py

```diff
@@ -140,6 +150,24 @@ def _on_reset_instance_action(self, event: ActionEvent) -> None:
         # results is a dict and set_results expects _SerializedData
         event.set_results(results)  # type: ignore[arg-type]
 
+    def _on_database_created(self, event: DatabaseCreatedEvent) -> None:
+        """Handle database created.
+
+        Args:
+            event: Event triggering the database created handler.
+        """
+        if not self.unit.is_leader():
+            event.defer()
+            return
+
+        self.model.unit.status = ops.MaintenanceStatus("Preparing the database")
+        try:
+            self._database.prepare_database()
+            self._on_config_changed(event)
+        except psycopg2.Error as exc:
+            self.model.unit.status = ops.BlockedStatus(str(exc))
```

Comment:
Can the user act in this situation ? If so can we put a more explicit message of what we expect them to do, if not I think we should put it in ErrorStatus

Reply:
Depending on the error, it might have some action to do on the database side, that's why I added it as blocked status

Reply:
This is quite an interesting case actually - it isn't clear whether the charm should be in blocked or error state. It also doesn't seem to be reasonable for the charm to have logic to figure out whats wrong since this is more on the database side.

Perhaps it would be good to have something in the database library from the data platform team where you can send an exception and set the charm state based on that? Any thoughts @benhoyt

Reply:
Yeah, I think Gregory's intuition is correct -- this should probably be error status. The error is basically "some problem with your database". However, instead of explicitly setting an error status, I'd just remove the try/except and let it fail. Juju will automatically put it in error status with a generic message (and the stack track will be in `juju debug-log`), and Juju will automatically retry.

It seems like we're catching any database error here, so we'd have no idea what message to put in a `BlockedStatus` anyway. It's not blocked (waiting for a config value or other user action), it's error (we have no idea what went wrong).

@amandahla What kind of errors were you seeing that you put a try/except here? Or was it just a theoretical concern?

Reply:
I didn't test but I was considering the scenario when a backup is being created and would block new connections. So the action would be to check the backup.
Or if for some reason the user has no permission to update the pg_database table then the action would be to check the table permissions or user roles.
Or even if for some reason the relation doesn't exist or it was removed.

I didn't think about ErrorStatus because of this "Error state that is never set from Charm code. " in [Constructs](https://juju.is/docs/sdk/constructs).

Anyway, IMHO it's a matter of interpretation about what is a problem that might need an action so I won't mind changing this for ErrorStatus.

Reply:
Yeah, you wouldn't set `ErrorStatus` explicitly, you'd just remove the `try/except` altogether and let the error bubble up to Juju, which would set an error status.

Reply:
Ok, try/except removed, thanks.